### PR TITLE
Fix zone balance

### PR DESCRIPTION
**Pull request by darionyaphet**

### CODE REVIEW

File: src/meta/processors/admin/Balancer.cpp [hunk 1]
```diff
@@ -389,6 +398,39 @@ bool Balancer::balanceParts(BalanceID balanceId,
 
   auto maxPartsHost = sortedHosts.back();
   auto minPartsHost = sortedHosts.front();
+  if (innerBalance_) {
```
Comment (critical27) [inline src/meta/processors/admin/Balancer.cpp:401]: What is `innerBalance_`?
What is the relationship with `update` function...

Comment (darionyaphet) [reply src/meta/processors/admin/Balancer.cpp:401]: `innerBalance_ ` is used to move parts in one zone. `update` is try to collect the modify into one place...

File: src/meta/processors/admin/Balancer.cpp [hunk 2]
```diff
@@ -490,50 +614,88 @@ ErrorOr<nebula::cpp2::ErrorCode, bool> Balancer::getHostParts(GraphSpaceID space
...
+    }
+    for (const auto& h : lostHosts) {
+      LOG(INFO) << "Lost host " << h;
+      confirmedHostParts.erase(h);
+    }
+
+    auto zonePartsRet = assembleZoneParts(groupName, confirmedHostParts);
+    if (zonePartsRet != nebula::cpp2::ErrorCode::SUCCEEDED) {
+      LOG(ERROR) << "Assemble Zone Parts failed group: " << groupName;
+      return zonePartsRet;
```
Comment (critical27) [inline src/meta/processors/admin/Balancer.cpp:662]: What is this? I guess you just wanna generate `innerBalance_`

**Comment by bright-starry-sky** [comment]
could you give some comments for PR purpose? Or related issue?

**Comment by darionyaphet** [comment]
> could you give some comments for PR purpose? Or related issue?

Actually, this PR is migrated from the nebula-storage which will fix the balance with zone test case.

File: src/meta/processors/admin/Balancer.cpp [hunk 3]
```diff
@@ -416,41 +462,125 @@ bool Balancer::balanceParts(BalanceID balanceId,
...
+
+      auto& source = maxPartsHost.first;
+      auto iter = std::find_if(zoneParts_.begin(), zoneParts_.end(),
+                               [&source](const auto& pair) {
+        return source == pair.first;
+      });
+
+      auto& zoneName = iter->second.first;
+      int32_t hostsSize = zoneHosts_[zoneName].size();
+      int32_t totalPartsZone = 0;
```
Comment (bright-starry-sky) [inline src/meta/processors/admin/Balancer.cpp:517]: iter != zoneParts_.end() ?

File: src/meta/processors/admin/Balancer.cpp [hunk 4]
```diff
@@ -416,41 +462,125 @@ bool Balancer::balanceParts(BalanceID balanceId,
...
+  LOG(INFO) << "Update min and max loading";
+  LOG(INFO) << "Total parts in zone " << totalPartsZone
+            << ", total hosts " << hostsSize;
+  avgLoad = static_cast<float>(totalPartsZone) / hostsSize;
+  LOG(INFO) << "The expect avg load is " << avgLoad;
+  minLoad = std::floor(avgLoad);
+  maxLoad = std::ceil(avgLoad);
+  LOG(INFO) << "The min load is " << minLoad << " max load is " << maxLoad;
+}
+
```
Comment (bright-starry-sky) [inline src/meta/processors/admin/Balancer.cpp:583]: What's purpose? I don't see it being called.

File: src/meta/processors/admin/Balancer.cpp [hunk 5]
```diff
@@ -554,15 +723,38 @@ nebula::cpp2::ErrorCode Balancer::assembleZoneParts(const std::string& groupName
...
+    auto& hosts = it->second;
+    for (auto hostIter = hosts.begin(); hostIter != hosts.end(); hostIter++) {
+      auto h = *hostIter;
+      auto iter = std::find_if(hostParts.begin(), hostParts.end(),
+                               [h](const auto& pair) -> bool {
+        return h == pair.first;
+      });
+
+      if (iter == hostParts.end()) {
+        LOG(INFO) << "";
```
Comment (bright-starry-sky) [inline src/meta/processors/admin/Balancer.cpp:749]: ?

**Comment by bright-starry-sky** [comment]
Ready to review ?

File: src/meta/processors/admin/Balancer.cpp (hunk 1 above)
Comment (darionyaphet) [reply src/meta/processors/admin/Balancer.cpp:402]: so I have remove temporarily

File: src/meta/processors/admin/Balancer.cpp (hunk 3 above)
Comment (darionyaphet) [reply src/meta/processors/admin/Balancer.cpp]: check in `checkZoneLegal`

File: src/meta/processors/admin/Balancer.cpp (hunk 2 above)
Comment (darionyaphet) [reply src/meta/processors/admin/Balancer.cpp:668]: `assembleZoneParts` collect some info about zone

**Review by liuyu85cn — APPROVED** [review]

File: src/meta/processors/admin/Balancer.cpp (hunk 2 above)
Comment (critical27) [reply src/meta/processors/admin/Balancer.cpp:662]: Explain what are you doing from 638 to 663. Why we `calDiff` and `assembleZoneParts`, nothing is returned.

**Comment by bright-starry-sky** [comment]
Please don't mark the conversation to resolved if there is no modification and reason.

File: src/meta/processors/admin/Balancer.cpp [hunk 6]
```diff
@@ -1052,13 +1239,29 @@ bool Balancer::checkZoneLegal(const HostAddr& source, const HostAddr& target, Pa
...
-  if (sourceIter->second.first == targetIter->second.first) {
-    LOG(INFO) << source << " --> " << target << " transfer in the same zone";
+  // if (sourceIter->second.first == targetIter->second.first) {
+  //   LOG(INFO) << source << " --> " << target << " transfer in the same zone";
+  //   return true;
+  // }
+
+  // auto& parts = targetIter->second.second;
+  // return std::find(parts.begin(), parts.end(), part) == parts.end();
+
```
Comment (bright-starry-sky) [inline src/meta/processors/admin/Balancer.cpp:1249]: could you delete the useless code?

File: src/meta/processors/admin/Balancer.cpp [hunk 7]
```diff
@@ -1052,13 +1239,29 @@ bool Balancer::checkZoneLegal(const HostAddr& source, const HostAddr& target, Pa
...
+  if (!innerBalance_) {
+    LOG(INFO) << "innerBalance_ is false";
     return true;
+  } else {
+    LOG(INFO) << "same zone";
+    LOG(INFO) << sourceIter->second.first << " : " << targetIter->second.first;
+    return sourceIter->second.first == targetIter->second.first;
```
Comment (bright-starry-sky) [inline src/meta/processors/admin/Balancer.cpp:1256]: A suggestion, this step can be move to top of method.

File: src/meta/processors/admin/Balancer.cpp [hunk 8]
```diff
@@ -424,33 +466,115 @@ bool Balancer::balanceParts(BalanceID balanceId,
...
+  int32_t hostsSize = zoneHosts_[zoneName].size();
+  int32_t totalPartsZone = 0;
+  for (auto& host : zoneHosts_[zoneName]) {
+    auto it = confirmedHostParts.find(host);
+    if (it == confirmedHostParts.end()) {
+      LOG(ERROR) << "Host " << host << "not in confirmedHostParts";
+      continue;
+    }
+    totalPartsZone += it->second.size();
+  }
```
Comment (bright-starry-sky) [inline src/meta/processors/admin/Balancer.cpp:567]: Please remove the useless code

**Comment by darionyaphet** [comment]
> Please don't mark the conversation to resolved if there is no modification and reason.

These places have been changed, but there is no push  orz .

File: src/meta/processors/admin/Balancer.cpp [hunk 9]
```diff
@@ -554,15 +717,37 @@ nebula::cpp2::ErrorCode Balancer::assembleZoneParts(const std::string& groupName
...
+      });
+
+      if (iter == hostParts.end()) {
+        LOG(INFO) << "";
+        continue;
+      }
+
+      auto& parts = iter->second;
+      auto& hp = relatedParts_[host];
+      hp.insert(hp.end(), parts.begin(), parts.end());
```
[the remark this answers is not in the document]
Comment (darionyaphet) [reply src/meta/processors/admin/Balancer.cpp:748]: zoneHosts use to record this host belong to zone's hosts.

it's a std::unordered_map<std::pair<HostAddr, std::string>, std::vector<HostAddr>>

File: src/meta/processors/admin/Balancer.cpp (hunk 2 above)
Comment (darionyaphet) [reply src/meta/processors/admin/Balancer.cpp:624]: if the space dependent on group, it use to record the partition contained in the zone related to the node.

`assembleZoneParts ` save it in relatedParts_.

calDiff save it in the argument which is reference .

File: src/meta/processors/admin/Balancer.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (bright-starry-sky) [reply src/meta/processors/admin/Balancer.cpp:709]: So I think the relatedParts_[host] = parts is enough. WDYT ? 
don't need an ordered insertion.

File: src/meta/processors/admin/Balancer.cpp [hunk 10]
```diff
@@ -424,30 +466,74 @@ bool Balancer::balanceParts(BalanceID balanceId,
...
     }
+
     if (noAction) {
       LOG(INFO) << "Here is no action";
       break;
     }
     sortedHosts = sortedHostsByParts(confirmedHostParts);
     maxPartsHost = sortedHosts.back();
     minPartsHost = sortedHosts.front();
+    if (innerBalance_) {
```
Comment (bright-starry-sky) [inline src/meta/processors/admin/Balancer.cpp:497]: What's innerBalance_ means?  means the zone size > replica size ?

Comment (bright-starry-sky) [reply src/meta/processors/admin/Balancer.cpp:497]: > What's innerBalance_ means? means the zone size > replica size ?

no, should be >=

File: src/meta/processors/admin/Balancer.cpp [hunk 11]
```diff
@@ -424,30 +466,74 @@ bool Balancer::balanceParts(BalanceID balanceId,
...
+      LOG(INFO) << "maxPartsHost.first " << maxPartsHost.first << " minPartsHost.first "
+                << minPartsHost.first;
```
Comment (bright-starry-sky) [inline src/meta/processors/admin/Balancer.cpp:499]: don't output the code block for log. WDYT?

File: src/meta/processors/admin/Balancer.cpp (hunk 10 above)
Comment (darionyaphet) [reply src/meta/processors/admin/Balancer.cpp:497]: `innerBalance_ = (replica == zoneSize)`  use to mark the replication number same with  zone size

File: src/meta/processors/admin/Balancer.cpp [hunk 12]
```diff
@@ -424,30 +466,74 @@ bool Balancer::balanceParts(BalanceID balanceId,
...
+        }
+        totalPartsZone += it->second.size();
+      }
+
+      LOG(INFO) << "Update min and max loading";
+      LOG(INFO) << "Total parts in zone " << totalPartsZone << ", total hosts " << hostsSize;
+      avgLoad = static_cast<float>(totalPartsZone) / hostsSize;
+      LOG(INFO) << "The expect avg load is " << avgLoad;
+      minLoad = std::floor(avgLoad);
+      maxLoad = std::ceil(avgLoad);
```
Comment (bright-starry-sky) [inline src/meta/processors/admin/Balancer.cpp:527]: From line 497 to 527 , these codes just for one LOG (line 528), right?

File: src/meta/processors/admin/Balancer.cpp (hunk 7 above)
Comment (bright-starry-sky) [reply src/meta/processors/admin/Balancer.cpp:1209]: > A suggestion, this step can be move to top of method.

Please ignore the comment. my fault.

File: src/meta/processors/admin/Balancer.cpp (hunk 10 above)
Comment (bright-starry-sky) [reply src/meta/processors/admin/Balancer.cpp:497]: > `innerBalance_ = (replica == zoneSize)` use to mark the replication number same with zone size

got

File: src/meta/processors/admin/Balancer.cpp (hunk 12 above)
Comment (darionyaphet) [reply src/meta/processors/admin/Balancer.cpp:522]: merge logs into one line

File: src/meta/processors/admin/Balancer.cpp (hunk 11 above)
Comment (darionyaphet) [reply src/meta/processors/admin/Balancer.cpp]: ok

**Review by liuyu85cn — APPROVED** [review]

**Review by panda-sheep — APPROVED** [review]

**Pull request merged**